src/prefect/new_task_engine.py

```diff
@@ -0,0 +1,382 @@
+        return not retry_condition or retry_condition(
+            self.task, self.task_run, self.state
+        )  # type: ignore
+
+    async def _run_hooks(self, state: State) -> None:
+        """Run the on_failure and on_completion hooks for a task, making sure to
+        catch and log any errors that occur.
+        """
+        task = self.task
+        task_run = self.task_run
+
+        hooks = None
+        if state.is_failed() and task.on_failure:
+            hooks = task.on_failure
+        elif state.is_completed() and task.on_completion:
+            hooks = task.on_completion
+
+        if hooks:
+            for hook in hooks:
+                hook_name = _get_hook_name(hook)
+                try:
+                    self.logger.info(
+                        f"Running hook {hook_name!r} in response to entering state"
+                        f" {state.name!r}"
+                    )
+                    if is_async_fn(hook):
+                        await hook(task=task, task_run=task_run, state=state)
+                    else:
+                        hook(task=task, task_run=task_run, state=state)
+                except Exception:
+                    self.logger.error(
+                        f"An error was encountered while running hook {hook_name!r}",
+                        exc_info=True,
+                    )
+                else:
+                    self.logger.info(
+                        f"Hook {hook_name!r} finished running successfully"
+                    )
+
+    def _compute_state_details(
+        self, include_cache_expiration: bool = False
+    ) -> StateDetails:
+        ## setup cache metadata
+        task_run_context = TaskRunContext.get()
+        cache_key = (
+            self.task.cache_key_fn(
+                task_run_context,
+                self.parameters,
+            )
+            if self.task.cache_key_fn
+            else None
+        )
+        # Ignore the cached results for a cache key, default = false
+        # Setting on task level overrules the Prefect setting (env var)
+        refresh_cache = (
+            self.task.refresh_cache
+            if self.task.refresh_cache is not None
+            else PREFECT_TASKS_REFRESH_CACHE.value()
+        )
+
+        if include_cache_expiration:
+            cache_expiration = (
+                (pendulum.now("utc") + self.task.cache_expiration)
+                if self.task.cache_expiration
+                else None
+            )
+        else:
+            cache_expiration = None
+        return StateDetails(
+            cache_key=cache_key,
+            refresh_cache=refresh_cache,
+            cache_expiration=cache_expiration,
+        )
+
+    async def begin_run(self) -> State:
+        state_details = self._compute_state_details()
+        new_state = Running(state_details=state_details)
+        state = await self.set_state(new_state)
+        while state.is_pending():
+            await asyncio.sleep(1)
```

Comment:
Why do we sleep here?

Reply:
just to prevent a tight while loop, open to revisit